compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64ArrayCompareToOp.java

```diff
@@ -120,177 +140,300 @@ protected void emitCode(CompilationResultBuilder crb, AArch64MacroAssembler masm
 
-        // ... or string less than vector length.
-        masm.bind(COMPARE_SHORT_LABEL);
-        for (int i = 0; i < VECTOR_SIZE_BYTES; i += CHAR_SIZE_BYTES) {
-            if (isLU || isUL) {
-                masm.ldr(8, temp, AArch64Address.createImmediateAddress(8, AArch64Address.AddressingMode.IMMEDIATE_SIGNED_UNSCALED, array1, i / 2));
-            } else {
-                int transferSize = 8 * CHAR_SIZE_BYTES;
-                masm.ldr(transferSize, temp, AArch64Address.createImmediateAddress(transferSize, AArch64Address.AddressingMode.IMMEDIATE_SIGNED_UNSCALED, array1, i));
-            }
+        int charSizeBytes = 1;
+        final int vectorSizeBytes = 32;
 
-            int transferSize = 8 * CHAR_SIZE_BYTES;
-            masm.ldr(transferSize, result, AArch64Address.createImmediateAddress(transferSize, AArch64Address.AddressingMode.IMMEDIATE_SIGNED_UNSCALED, array2, i));
+        /* Byte is expanded to short if we compare non-LL strings. */
+        if (!isLL) {
+            charSizeBytes = 2;
+        }
 
-            if (isUL) {
-                // UL's input has been swapped in AArch64StringUTF16Substitutions.compareToLatin1.
-                masm.subs(64, result, result, temp);
-            } else {
-                masm.subs(64, result, temp, result);
-            }
+        masm.mov(64, vecCount, zr);
+        /* tail count (in bytes) */
+        masm.and(64, tailCount, length, vectorSizeBytes - 1);
 
-            masm.branchConditionally(ConditionFlag.NE, BREAK_LABEL);
-            masm.subs(64, tailCount, tailCount, CHAR_SIZE_BYTES);
-            masm.branchConditionally(ConditionFlag.EQ, LENGTH_DIFFER_LABEL);
+        masm.align(16);
+        masm.bind(charSearchLoop);
+        int transferSize = 8 * charSizeBytes;
+        if (isLU || isUL) {
+            masm.ldr(8, temp, AArch64Address.createImmediateAddress(8, AArch64Address.AddressingMode.IMMEDIATE_POST_INDEXED, array1, 1));
+        } else {
+            masm.ldr(transferSize, temp, AArch64Address.createImmediateAddress(transferSize, AArch64Address.AddressingMode.IMMEDIATE_POST_INDEXED, array1, charSizeBytes));
         }
+        masm.ldr(transferSize, result, AArch64Address.createImmediateAddress(transferSize, AArch64Address.AddressingMode.IMMEDIATE_POST_INDEXED, array2, charSizeBytes));
+        if (isUL) {
+            /* UL's input has been swapped in AArch64StringUTF16Substitutions.compareToLatin1. */
+            masm.subs(64, result, result, temp);
+        } else {
+            masm.subs(64, result, temp, result);
+        }
+        masm.branchConditionally(ConditionFlag.NE, breakLabel);
+        masm.subs(64, tailCount, tailCount, charSizeBytes);
+        masm.branchConditionally(ConditionFlag.EQ, stringsEqualUptoLength);
+        masm.jmp(charSearchLoop);
 
-        // STRING_DIFFER extract exact value of a difference.
-        masm.bind(STRING_DIFFER_LABEL);
+        /* Extract exact value of a difference. */
+        masm.bind(stringDifferLabel);
```

Comment:
Unless I'm missing something, I don't think this code is reachable - this label is never targeted and there is an unconditional jump immediately above it